pontoon/static/js/widgets/table.js

```diff
@@ -0,0 +1,153 @@
+
+import React from 'react';
+import ReactTable from 'react-table';
+import "react-table/react-table.css";
+
+import Checkbox from 'widgets/checkbox';
+
+
+export class CheckboxTable extends React.Component {
+
+    constructor (props) {
+        super(props);
+        this.visible = [];
+        this.state = {checked: new Set()};
+        this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
+        this.handleSelectAll = this.handleSelectAll.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+        this.handleTableChange = this.handleTableChange.bind(this);
+        this.handleTableSortChange = this.handleTableSortChange.bind(this);
+        this.handleTableResize = this.handleTableResize.bind(this);
+        this.renderCheckbox = this.renderCheckbox.bind(this);
+        this.renderSelectAllCheckbox = this.renderSelectAllCheckbox.bind(this);
+    }
+
+    get columnSelect () {
+        return {
+            Header: this.renderSelectAllCheckbox,
+            Cell: this.renderCheckbox,
+            sortable: false,
+            width: 45};
+    }
+
+    get defaultPageSize () {
+        return 5;
+    }
+
+    get pruned() {
+        // Returns a copy of the checked set with any resource paths that are
+        // not in `this.visible` removed
+        let {checked} = this.state;
+        return new Set([...checked].filter(v => (this.visible.indexOf(v) !== -1)))
+    }
+
+    get selected () {
+        // get the pruned list of selected resources, returns all=true/false
+        // if all visible resources are selected
+        // some rows can be empty strings if there are more visible rows than
+        // resources
+        const checked = this.pruned
+        const all = (
+            checked.size > 0
+                && checked.size === this.visible.filter(v => v !== '').length);
+        return {all, checked};
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            this.visible.length = 0;
+            this.setState({checked: new Set()});
+        }
+    }
+
+    handleCheckboxClick (evt) {
+        // adds/removes paths for submission
+        let {checked} = this.state;
+        checked = new Set(checked);
+        evt.target.checked ? checked.add(evt.target.name) : checked.delete(evt.target.name);
+        this.setState({checked});
```

Comment:
Note how setting the new state depending on the current value of `state` should be done [via the function argument](https://reactjs.org/docs/faq-state.html#how-do-i-update-state-with-values-that-depend-on-the-current-state).